src/server/scheduler.rs

```diff
@@ -0,0 +1,406 @@
+    },
+    clamp::Clamp,
+    effect_duration::EffectDuration,
+};
+
+use crate::{backend::BackendRequest, error::ApiResult, resource::Resources};
+
+#[derive(Debug)]
+pub struct Scheduler {
+    jobs: Vec<JoinHandle<()>>,
+    res: Arc<Mutex<Resources>>,
+    behavior_instances: Vec<ScheduleBehaviorInstance>,
+}
+
+impl Scheduler {
+    pub const fn new(res: Arc<Mutex<Resources>>) -> Self {
+        Self {
+            jobs: vec![],
+            behavior_instances: vec![],
+            res,
+        }
+    }
+
+    pub async fn update(&mut self) {
+        let new_behavior_instances = self.get_behavior_instances().await;
+        if new_behavior_instances != self.behavior_instances {
+            self.behavior_instances = new_behavior_instances;
+            self.update_jobs();
+        }
+    }
+
+    fn update_jobs(&mut self) {
+        for job in &self.jobs {
+            job.abort();
+        }
+        self.jobs = self
+            .behavior_instances
+            .iter()
+            .filter(|ScheduleBehaviorInstance(_, bi)| bi.enabled)
+            .flat_map(|ScheduleBehaviorInstance(id, bi)| match &bi.configuration {
+                BehaviorInstanceConfiguration::Wakeup(wakeup_configuration) => {
+                    wakeup(self.res.clone(), id, wakeup_configuration)
+                }
+            })
+            .collect();
+    }
+
+    async fn get_behavior_instances(&self) -> Vec<ScheduleBehaviorInstance> {
+        self.res
+            .lock()
+            .await
+            .get_resources_by_type(RType::BehaviorInstance)
+            .into_iter()
+            .filter_map(|r| match r.obj {
+                Resource::BehaviorInstance(behavior_instance) => {
+                    Some(ScheduleBehaviorInstance(r.id, behavior_instance))
+                }
+                _ => None,
+            })
+            .collect()
+    }
+}
+
+fn wakeup(
+    res: Arc<Mutex<Resources>>,
+    id: &Uuid,
+    wakeup_configuration: &WakeupConfiguration,
+) -> Vec<JoinHandle<()>> {
+    let jobs = create_wake_up_jobs(id, wakeup_configuration);
+    jobs.into_iter()
+        .map(move |job| spawn(job.create(res.clone())))
+        .collect()
+}
+
+async fn disable_behavior_instance(id: Uuid, res: Arc<Mutex<Resources>>) {
+    let upd = BehaviorInstanceUpdate::default().with_enabled(false);
+    let upd_result = res
+        .lock()
+        .await
+        .update::<BehaviorInstance>(&id, |bi| *bi += upd);
+    if let Err(err) = upd_result {
+        log::error!("Failed to disable behavior instance {:?}", err);
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct ScheduleBehaviorInstance(Uuid, BehaviorInstance);
+
+#[derive(Debug)]
+enum ScheduleType {
+    Recurring(Weekday),
+    Once(),
+}
+
+pub struct WakeupJob {
+    resource_id: Uuid,
+    schedule_type: ScheduleType,
+    configuration: WakeupConfiguration,
+}
+
+impl WakeupJob {
+    fn start_datetime(&self, now: DateTime<Local>) -> Result<DateTime<Local>, &'static str> {
```

Comment:
An error type of `&'static str`, although convenient, is not very good.

Feel free to introduce either a local error enum, or new error variants to `ApiError`, which (yes, despite the name.. :sweat_smile:) is broadly used in Bifrost.

Reply:
Yeah, forgot to fix this. I used the string values while debugging, but it's probably nice to have proper errors if other users experiences errors in this part of the code.

Reply:
Definitely. Just use `ApiResult<>` while in the main Bifrost crate :+1: 

(feel free to introduce new error variants if you feel it's needed)